# 1. Importing Libraries

In [2]:
# import libraries
import pandas as pd
import numpy as np
import os

# 2. Importing Data

In [3]:
# This command will import my saved CSV files within a certain folder
df = pd.read_csv(r'/Users/obinwannealisigwe/Desktop/Python/Achievement_4_Project_Obi/02. Data/Original Data/orders.csv', index_col=False)

In [4]:
# This command tells python to remember a main folder path
path_1 = r'/Users/obinwannealisigwe/Desktop/Python/Achievement_4_Project_Obi/02. Data/Original Data'

In [5]:
# This command tells python to remember a main folder path
path = r'/Users/obinwannealisigwe/Desktop/Python/Achievement_4_Project_Obi/02. Data/Prepared Data'

In [6]:
path

'/Users/obinwannealisigwe/Desktop/Python/Achievement_4_Project_Obi/02. Data/Prepared Data'

In [7]:
# Import the “products.csv” file into Jupyter
df_prod = pd.read_csv(os.path.join(path_1, 'products.csv'), index_col = False)

In [8]:
# Import the “orders_wrangled.csv” file into Jupyter
df_ord = pd.read_csv(os.path.join(path, 'orders_wrangled.csv'), index_col = False)

In [9]:
df_ord.head()

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0           0   2539329        1             1                   2   
1           1   2398795        1             2                   3   
2           2    473747        1             3                   3   
3           3   2254736        1             4                   4   
4           4    431534        1             5                   4   

   order_hour_of_day  days_since_prior_order  
0                  8                     NaN  
1                  7                    15.0  
2                 12                    21.0  
3                  7                    29.0  
4                 15                    28.0

# 3. Data Consistency Checks

In [10]:
df_ord.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_prior_order  
count        3.421083e+06       3.421083e+06            3.214874e+06  
mean         2.776219e+00       1.345202e+01            1.111484e+01  
std          2.046829e+00       4.226088e+00            9.206737e+00  
min          0.000000e+00       0.000000e+00            0.000000e+00  
25%          1.000000e+00       1.000000e+01            4.000000e+00  
50%          3.000000e+00       1.300000e+01            7.000000e+00  
75%          5.000000e+00       1.600000e+01            1.500000e+01  
max          6.000000e+00       2.300000e+01            3.000000e+01

## Mixed-Type Data

In [11]:
# Create a dataframe
df_test = pd.DataFrame()

In [12]:
# Create a mixed type column
df_test['mix'] = ['a', 'b', 1, True]

In [13]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [14]:
# check for mixed types
for col in df_test.columns.tolist():
  weird = (df_test[[col]].map(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [15]:
# fix mixed-type column
df_test['mix'] = df_test['mix'].astype('str')

## Missing Values

In [16]:
df_prod.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [17]:
df_nan = df_prod[df_prod['product_name'].isnull() == True]

In [18]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [19]:
df_prod.shape

(49693, 5)

In [20]:
# create a subset containing only non-missing values
df_prod_clean = df_prod[df_prod['product_name'].isnull() == False]

In [21]:
df_prod_clean.shape

(49677, 5)

In [22]:
# create a subset containing only missing values
df_prod.dropna(subset = ['product_name'], inplace = True)

In [23]:
df_prod.isnull().sum()

product_id       0
product_name     0
aisle_id         0
department_id    0
prices           0
dtype: int64

## Duplicate Values

In [24]:
df_dups = df_prod_clean[df_prod_clean.duplicated()]

In [25]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [26]:
# check current number of rows in df_prod_clean
df_prod_clean.shape

(49677, 5)

In [27]:
# creating dataframe tha doesn't include duplicates
df_prod_clean_no_dups = df_prod_clean.drop_duplicates()

In [28]:
df_prod_clean_no_dups.shape

(49672, 5)

# Exercise 4.5 Task

## 2. Run the df.describe() function on your df_ords dataframe.

In [29]:
# 2. check the df_ord dataframe for inconsistencies
df_ord.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_prior_order  
count        3.421083e+06       3.421083e+06            3.214874e+06  
mean         2.776219e+00       1.345202e+01            1.111484e+01  
std          2.046829e+00       4.226088e+00            9.206737e+00  
min          0.000000e+00       0.000000e+00            0.000000e+00  
25%          1.000000e+00       1.000000e+01            4.000000e+00  
50%          3.000000e+00       1.300000e+01            7.000000e+00  
75%          5.000000e+00       1.600000e+01            1.500000e+01  
max          6.000000e+00       2.300000e+01            3.000000e+01

### 'days_since_prior_order' has min value as 0, and max value as 30, which suggest that orders are capped montly

### 'order_number has a max value of 100, which might be too high depending on the time frame'

### 'order_day_of_week' has min value of 0 and max value of 6 meaning the 7 number (0-6) represent the day of the week

### 'order_hour_of_day' has min value of 0 and max value of 23 meaning the 24hrs (0-23) represent 23hrs of the day

## 3. Check for mixed-type data in your df_ord dataframe

In [30]:
# check for mixed-txpe data in df_ord dataframe
for col in df_ord.columns.tolist():
  weird = (df_ord[[col]].map(type) != df_ord[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ord[weird]) > 0:
    print (col)

In [31]:
df_ord.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3421083 entries, 0 to 3421082
Data columns (total 7 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   Unnamed: 0              int64  
 1   order_id                int64  
 2   user_id                 int64  
 3   order_number            int64  
 4   orders_day_of_week      int64  
 5   order_hour_of_day       int64  
 6   days_since_prior_order  float64
dtypes: float64(1), int64(6)
memory usage: 182.7 MB


### No mixed-type data was found

## 5. Run a check for missing values in your df_ords dataframe

In [32]:
# find missing values
df_ord.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

### The 'days_since_prior_order' column is missing 206209 values
### This could be that the customer first order don't have a value
### I woundn't suggest that we remove this it instead we will preserve it

# 6. Addressing the missing values

In [33]:
# Create a subset containing only missing values
df_nan = df_ord[df_ord['days_since_prior_order'].isnull() == True]

In [34]:
df_nan

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0                 0   2539329        1             1                   2   
11               11   2168274        2             1                   2   
26               26   1374495        3             1                   1   
39               39   3343014        4             1                   6   
45               45   2717275        5             1                   3   
...             ...       ...      ...           ...                 ...   
3420930     3420930    969311   206205             1                   4   
3420934     3420934   3189322   206206             1                   3   
3421002     3421002   2166133   206207             1                   6   
3421019     3421019   2227043   206208             1                   1   
3421069     3421069   3154581   206209             1                   3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 7 columns]

### Lets verify if all 'order_number' values = 1 are missing values in the 'days_since_prior_order' column

In [35]:
# To determine the action to take regarding the missing values, lets separate the data from the original dataframe
df_ord_nan = df_nan[df_nan['days_since_prior_order'].isnull() == True]

In [36]:
df_nan[df_nan['order_number'] == '1']

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

### # Letsverify if all 'order_number' values equal to 1 are missing values in the 'days_since_prior_order' column

In [37]:
# Filter the dataframe where 'order_number' equals 1
df_first_orders = df_ord_nan[df_ord_nan['order_number'] == '1']

In [38]:
df_nan[(df_nan['order_number'] == '1')]

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

In [39]:
# Check if all of these have missing values in 'days_since_prior_order'
df_first_orders_missing = df_first_orders['days_since_prior_order'].isna().all()

In [40]:
df_first_orders['days_since_prior_order']

Series([], Name: days_since_prior_order, dtype: float64)

In [41]:
# Print the result
print("All first orders have missing 'days_since_prior_order':", df_first_orders_missing) 

All first orders have missing 'days_since_prior_order': True


### We confirm that every row with 'order_number' equal to 1 has a missing value in the'days_since_prior_order' column

In [42]:
# Address the missing values by creating a column that flags order numbers as either True/False depending on if the 'order_number' = 1
df_ord['first_order'] = df_ord['days_since_prior_order'].isna()

In [43]:
df_ord

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0                 0   2539329        1             1                   2   
1                 1   2398795        1             2                   3   
2                 2    473747        1             3                   3   
3                 3   2254736        1             4                   4   
4                 4    431534        1             5                   4   
...             ...       ...      ...           ...                 ...   
3421078     3421078   2266710   206209            10                   5   
3421079     3421079   1854736   206209            11                   4   
3421080     3421080    626363   206209            12                   1   
3421081     3421081   2977660   206209            13                   1   
3421082     3421082    272231   206209            14                   6   

         order_hour_of_day  days_since_prior_order  first_order  
0                        8                     NaN         True  
1                        7                    15.0        False  
2                       12                    21.0        False  
3                        7                    29.0        False  
4                       15                    28.0        False  
...                    ...                     ...          ...  
3421078                 18                    29.0        False  
3421079                 10                    30.0        False  
3421080                 12                    18.0        False  
3421081                 12                     7.0        False  
3421082                 14                    30.0        False  

[3421083 rows x 8 columns]

In [43]:
# Check data-type of new column and update if necessary
df_ord['first_order'].dtype

dtype('bool')

In [44]:
# Re-check data to make sure totals are accurate
df_ord.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
first_order                    0
dtype: int64

### Missing values have been addressed appropriately, now the client can see why there are missing values in the 'order_number' column

### Removing/filtering these values would've been incorrect because the missing data is quite valuable 

# 7. Run a check for duplicate values in your df_ords data

In [45]:
df_ord_dup = df_ord[df_ord.duplicated()]

In [46]:
df_ord_dup

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order, first_order]
Index: []

### No Duplicates Found

# 9. Exporting Dataframes

In [47]:
path_2 = r'/Users/obinwannealisigwe/Desktop/Python/Achievement_4_Project_Obi/02. Data/Prepared Data'

In [48]:
# Exporting df_prods_clean_no_dups dataframe as "products_checked.csv"
df_prod_clean_no_dups.to_csv(os.path.join(path_2,'products_checked.csv'))

In [49]:
# Exporting df_ord dataframe as "orders_checked.csv"
df_ord.to_csv(os.path.join(path_2,'orders_checked.csv'))